Contributing to Actual Budget (#176)
actualbudget/docs: docs/Developers/contributing.md

---
title: Contributing to Actual Budget
---

> So, you want to get stuck in an help out with existing issues in Actual Budget or develop a feature of your own, that is great and we really appreciate that. 

Reviewer: ```suggestion
So, you want to get stuck in and help out with existing issues in Actual Budget or develop a feature of your own. That’s great and we really appreciate it!
```

### Main Contributors

(sorted alphabetically)

* [albertogasparin]
* [j-f1]
* [jlongster]
* [MatissJanis]
* [rich-howell]
* [trevdor]

### The Project Layout

The layout of the codebase in Actual Budget takes a bit of getting used to and finding things at first can be a little tricky, we have put together a help [document](/Developers/project-layout) that shows the structure of the project, while this isn't 100% complete it will give you a good starting point for your development. 

### Working on existing issues

Existing issues are a good place to start especially if you want to contribute to Actual Budget but don't know where to start, some of the things to be aware of are;

1. All issues are open to be worked on by anyone
2. Working on the highest rated [feature requests](https://github.com/actualbudget/actual/issues?q=label%3A%22needs+votes%22+sort%3Areactions-%2B1-desc+) would also be appreciated.
3. We do not assign issues to specific people

### Submitting an idea for something you want to work on

Okay, so you have an idea for something that you think would be great in Actual Budget, but how do you pitch it to the community so that all your hard work is not wasted?

> To start submit a pull request with the implementation or if it's a larger feature - open a new issue so we can discuss it.

Reviewer: ```suggestion
If the feature is relatively small, feel free to start the discussion by opening a PR. However, if you want to work on a larger change/feature, please open an issue or comment on an existing issue for the feature first. This lets the maintainers make sure your approach fits well both with the technical and ideological architecture of the project.
```

(feel free to change the words around, I’m not 100% happy with them)

### Writing Good Release Notes

> Create a Markdown file in the upcoming-release-notes directory of the repository you’re contributing to named after the PR number. The file should contain front matter with a category key (defining which header to put the entry under) and an authors key (defining the author of the entry). The body of the file should contain the changelog entry. Keep it short and clear — ideally one sentence, and also non-technical (unless the category is “Maintenance”). Copy-paste the template below to get started!

Reviewer: Is this moving out of the readme? If so it would be nice to delete it out of the readme so there isn’t duplication, then update the other repos to link here instead of there.

Author: Yes, it would make sense to put it in the docs rather than have bit of information in different readme's, I raised this https://github.com/actualbudget/actual-server/pull/212#issuecomment-1581268739

Try to phrase your message as a command, e.g. "Add option to include exchange rate multiplier during import" rather than "Added option to include exchange rate multiplier during import” or "Adds option to include exchange rate multiplier during import." Generally your message should match the PR title, but you can change it if you think it’s more clear.

### The Design Strategy Of Actual

The goal of the UI is to be minimalistic, but expose more advanced features progressively as the user interacts with the product (for example: the notes button is not visible by default if an account has no notes, but it becomes persistent visible if there are notes). We advocate for a similar approach in other places too. We are against adding a button/user setting for every little piece of UI (sizes, paddings, margins, etc.) as that goes against this simple design philosophy. 

The settings screen needs to also remain a place where core settings lives, we don't really want to have a myriad of options in here for each and every setting within the UI, doing that makes the code un-managable for future contributors and clutters up and confuses things for the users of Actual Budget. 

> ### Examples of pull request that will get rejected 

Reviewer: I kinda want to phrase this as something along the lines of “Things to avoid when opening a PR” so that it’s not negative (and also I don’t think it would be appropriate to name&shame real PRs we didn’t merge)

Author: Changed the heading here. This was never going to be a call out to closed PR's that had been rejected :) 

Would need some examples though of things to avoid.

[m3nu]: https://github.com/m3nu
[manuelcanepa]: https://github.com/manuelcanepa
[MatissJanis]: https://github.com/MatissJanis
[mnsrv]: https://github.com/mnsrv
[modrzew]: https://github.com/modrzew
[n1thun]: https://github.com/n1thun
[ostat]: https://github.com/ostat
[PartyLich]: https://github.com/PartyLich
[pmamberti]: https://github.com/pmamberti
[pole95]: https://github.com/pole95
[rianmcguire]: https://github.com/rianmcguire
[rich-howell]: https://github.com/rich-howell
[rickdoesdev]: https://github.com/rickdoesdev
[S3B4S]: https://github.com/S3B4S
[shall0pass]: https://github.com/shall0pass
[Silvenga]: https://github.com/Silvenga
[suryaatevellore]: https://github.com/suryaatevellore
[TomAFrench]: https://github.com/TomAFrench
[trevdor]: https://github.com/trevdor
[UnexomWid]: https://github.com/UnexomWid
[venkata-krishnas]: https://github.com/venkata-krishnas
[vincentscode]: https://github.com/vincentscode
[waseem-h]: https://github.com/waseem-h
[winklevos]: https://github.com/winklevos
[wmertens]: https://github.com/wmertens
[sinistersnare]: https://github.com/sinistersnare 
[henrikmaa]: https://github.com/henrikmaa 
[youngcw]: https://github.com/youngcw 

> [aaroneiche]: https://github.com/aaroneiche

Reviewer: Where are these coming from?

Author: Copied from release notes, pruned.

---

> title: Contributing to Actual Budget

Reviewer 2: Some more content in here: https://github.com/actualbudget/actual/blob/master/CONTRIBUTING.md

Also: once we have this page in docs - we can link the CONTRIBUTING.md doc to the new docs page.

Author: Took some bits from here and added. The setting up the Development env should really have it's own place in the docs here under the Developer section.

Reviewer 2: Great, thanks!

---

> 

Reviewer 2: Would be good to have a big warning: DO NOT EXPECT ALL PRs TO BE MERGED. In order for the project to stay healthy, we need to guard our bandwidth and also only take changes that align with Actual. Read more about the vision in this doc (+link to https://github.com/actualbudget/docs/pull/175).

Author: No, I am not putting a big warning in, we can had a paragraph about why that might not happen but having a massive warning saying DO NOT etc. seems unwelcoming.

Author: This is actually called out in the expectations 

> An open PR does not automatically deserve time for a full review and acceptance. It's up to the PR author to convince the maintainers that the change is good and worth reviewing. This involves a clear description for why the the change is being made, detailing the tradeoffs.

So, you want to get stuck in an help out with existing issues in Actual Budget or develop a feature of your own, that is great and we really appreciate that. 
We have created this document to signpost you to some of the key areas that will be of interest when developing for Actual Budget. 

As always, if you need any help or want something clarified, jump into the Discord and we will try our best to help you out. 

### The Project Layout

> The layout of the codebase in Actual Budget takes a bit of getting used to and finding things at first can be a little tricky, we have put together a help document that shows the structure of the project, while this isn't 100% complete it will give you a good starting point for your development. 

Reviewer 2: Where's the help document?

Author: I have added this in

### Working on existing issues

Existing issues are a good place to start especially if you are a first time contributor to Actual Budget, we have some issues labelled `good first issue` which are perfect for people who want to have a go at contributing. 

> If you are more familiar with Actual Budget, the `help wanted` label should point you in the right direction of some of the issues we need help with. 

Reviewer 2: I think we've basically gotten rid of these labels with the new feature management process we have. So instead of calling out specific labels, I'd recommend calling out that:
1. all bugs are open to be worked on by anyone
2. highest rated feature requests would also be appreciated (needs a link to the sorted list of feature request issues)
3. we do not assign issues to specific people

Author: The contributing.md document calls out labels https://github.com/actualbudget/actual/blob/master/CONTRIBUTING.md which I assume is where I got this from. 

Removed and replaced with your bullets.

Reviewer 2: Amazing, thanks!

---
category: Features
authors: [YourGitHubUsername]
---

Add option to include exchange rate multiplier during import
Valid categories:

* `Features`: New features
* `Enhancements`: Improvements to existing features
* `Bugfix`: Bug fixes
* `Maintenance`: Internal changes that don’t directly affect users
The `authors` key should be an array with the GitHub usernames of the people who contributed to the PR. In most cases, this should just be you but you can add multiple people if needed.

Try to phrase your message as a command, e.g. "Add option to include exchange rate multiplier during import" rather than "Added option to include exchange rate multiplier during import” or "Adds option to include exchange rate multiplier during import." Generally your message should match the PR title, but you can change it if you think it’s more clear.

### The Design Strategy Of Actual

> Actual Budget was designed to be sleek, clean and clutter free, by default there is a lot of empty space, this is by design. When developing your features or fixes take into consideration existing workflows such as drop down menus, submenus don't go straight for the empty space as that is more likely to get your code rejected. 

Reviewer 2: > don't go straight for the empty space as that is more likely to get your code rejected

Not sure I agree with this statement.

The goal of the UI is to be minimalistic, but expose more advanced features progressively as the user interacts with the product (for example: the notes button is not visible by default if an account has no notes, but it becomes persistent visible if there are notes). We advocate for a similar approach in other places too. We are against adding a button/user setting for every little piece of UI (sizes, paddings, margins, etc.) as that goes against this simple design philosophy.

Author: Maybe come up with something that suits then? We can add that in.

Reviewer 2: What do you think about the 2nd paragraph in my message? Would that be a good replacement here?

Author: > The goal of the UI is to be minimalistic, but expose more advanced features progressively as the user interacts with the product (for example: the notes button is not visible by default if an account has no notes, but it becomes persistent visible if there are notes). We advocate for a similar approach in other places too. We are against adding a button/user setting for every little piece of UI (sizes, paddings, margins, etc.) as that goes against this simple design philosophy.

Updated with your words
